src/launch/service.c

```diff
@@ -539,11 +540,104 @@ static int service_start_unit_handler(sd_bus_message *message, void *userdata, s
+                v = *(attrs++);
+
+                if (!k || !v)
+                        continue;
+
+                if (strcmp(k, "name") == 0 && strcmp(v, "StartUnitWithFlags") == 0) {
+                        *supported = 1;
+                        return;
+                }
+        }
+}
+
+static int start_methods_with_flags_supported(Launcher *launcher, bool *ret_supported) {
+        _c_cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _c_cleanup_(sd_bus_message_unrefp) sd_bus_message *reply_xml = NULL;
+        static int supported = -1;
+        XML_Parser parser;
+        const char *xml;
+        int r;
+
+        c_assert(launcher);
+        c_assert(ret_supported);
+
+        if (supported >= 0) {
+                *ret_supported = !!supported;
+                return 0;
+        }
+
+        r = sd_bus_call_method(launcher->bus_regular,
+                               "org.freedesktop.systemd1",
+                               "/org/freedesktop/systemd1",
+                               "org.freedesktop.DBus.Introspectable",
+                               "Introspect",
+                               &error, &reply_xml, NULL);
+        if (r < 0)
+                return error_origin(r);
+
+        r = sd_bus_message_read(reply_xml, "s", &xml);
+        if (r < 0)
+                return error_origin(r);
+
+        parser = XML_ParserCreate(NULL);
+        if (!parser)
+                return error_origin(-ENOMEM);
+
+        XML_SetUserData(parser, &supported);
+        XML_SetElementHandler(parser, parser_begin_fn, NULL);
+
+        supported = 0;
+
+        r = XML_Parse(parser, xml, strlen(xml), 1);
+        if (r != XML_STATUS_OK) {
+                log_append_here(&launcher->log, LOG_WARNING, 0, NULL);
+                r = log_commitf(&launcher->log,
+                                "Parsing the Introspect XML of /org/freedesktop/systemd1 failed, ignoring: %s\n",
+                                XML_ErrorString(XML_GetErrorCode(parser)));
+                if (r)
+                        return error_fold(r);
+
+                return -EINVAL;
+        }
+
+        *ret_supported = !!supported;
+
+        return 0;
+}
+
 static int service_start_unit(Service *service) {
         Launcher *launcher = service->launcher;
         _c_cleanup_(sd_bus_message_unrefp) sd_bus_message *method_call = NULL;
+        bool use_new_interface = false;
         int r;
 
+        /*
+         * When a Condition*= setting fails, the JobRemoved signal normally returns
+         * 'done' which we interpret as a success, but it means we will wait in vain
+         * for the service to start. It cannot be changed without breaking backward
+         * compatibility, so a new StartUnitWithFlags was added to change it on-demand.
+         * Let's use it if it is available, and fallback to the old StartUnit otherwise.
+         */
+        r = start_methods_with_flags_supported(launcher, &use_new_interface);
```

Comment:
Would it make sense to just try with the new interface, and then fallback to the old one if it fails (with the right error code)? We'd only try once of course, and remember the decision.

Reply:
I thought about that, but given it's all async, you wouldn't get a response immediately, so handling it properly you'd need to queue back the service start request from the response handler as a special case, and a lot of extra effort in general. Given dbus supports introspection exactly for this reason, and there's already an xml library being used in the project, it seemed to me to be much easier to go down this route. It's only done once and cached.

If the xml parsing is the problem, I could "hack" it a bit given we just need to know if the method is present, so in theory a simple check of the response xml string to see if it contains the "StartUnitWithFlags" substring would be enough to know, without full XML parsing. Would that be better?

Reply:
Yeah, that's a fair point. And if we are going to introspect, then we should absolutely do the full XML parsing as you are doing (though maybe it would be cleaner to have the code in a separate file, the same way we deal with loading the policy?).

I don't see any deadlock issues in this implementation, though my gut feeling is that blocking calls while handling async calls is a bit scary, and if we add more of these it might get convoluted quickly.

Would it be more predictable to move this into `launcher_run()`, maybe introducing a new helper function `launcher_introspect()` which could be extended in the future to read out more values, but for now just sets a boolean in the `Launcher` object telling us which interface to use?

Reply:
Yeah, I can look into moving it at startup, sounds like a good idea, will do so tomorrow

Reply:
Ok, updated as requested. Ptal!